File: src-tauri/src/commands/entity_pages.rs
use serde::{Deserialize, Serialize};
use tauri::State;
use uuid::Uuid;
use chrono::Utc;

use super::{AppState, CommandResult};

#[derive(Debug, Serialize, Deserialize)]
pub struct EntityPage {
    pub id: String,
    pub project_id: String,
    pub entity_id: String,
    pub entity_type: String,
    pub title: String,
    pub storage_path: String,
    pub metadata: Option<String>,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Deserialize)]
pub struct CreateEntityPageInput {
    pub project_id: String,
    pub entity_id: String,
    pub entity_type: String,
    pub title: String,
    pub storage_path: String,
    pub metadata: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct UpdateEntityPageInput {
    pub title: Option<String>,
    pub storage_path: Option<String>,
    pub metadata: Option<String>,
}

#[tauri::command]
pub fn create_entity_page(
    input: CreateEntityPageInput,
    state: State<AppState>,
) -> CommandResult<EntityPage> {
    let db = state.db.lock().unwrap();
    let id = Uuid::new_v4().to_string();
    let now = Utc::now().to_rfc3339();

    db.execute(
        "INSERT INTO entity_pages (
            id, project_id, entity_id, entity_type, title, storage_path,
            metadata, created_at, updated_at
        ) VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8, ?9)",
        [
            &id,
            &input.project_id,
            &input.entity_id,
            &input.entity_type,
            &input.title,
            &input.storage_path,
            &input.metadata.clone().unwrap_or_default(),
            &now,
            &now,
        ],
    )?;

    Ok(EntityPage {
        id,
        project_id: input.project_id,
        entity_id: input.entity_id,
        entity_type: input.entity_type,
        title: input.title,
        storage_path: input.storage_path,
        metadata: input.metadata,
        created_at: now.clone(),
        updated_at: now,
    })
}

#[tauri::command]
pub fn get_entity_page(
    entity_id: String,
    state: State<AppState>,
) -> CommandResult<Option<EntityPage>> {
    let db = state.db.lock().unwrap();

    let mut stmt = db.prepare(
        "SELECT id, project_id, entity_id, entity_type, title, storage_path,
        metadata, created_at, updated_at
        FROM entity_pages WHERE entity_id = ?1"
    )?;

    let result = stmt.query_row([&entity_id], |row| {
        Ok(EntityPage {
            id: row.get(0)?,
            project_id: row.get(1)?,
            entity_id: row.get(2)?,
            entity_type: row.get(3)?,
            title: row.get(4)?,
            storage_path: row.get(5)?,
            metadata: row.get(6)?,
            created_at: row.get(7)?,
            updated_at: row.get(8)?,
        })
    });

    match result {
        Ok(page) => Ok(Some(page)),
        Err(rusqlite::Error::QueryReturnedNoRows) => Ok(None),
        Err(e) => Err(e.into()),
    }
}

#[tauri::command]
pub fn list_entity_pages(
    project_id: String,
    state: State<AppState>,
) -> CommandResult<Vec<EntityPage>> {
    let db = state.db.lock().unwrap();

    let mut stmt = db.prepare(
        "SELECT id, project_id, entity_id, entity_type, title, storage_path,
        metadata, created_at, updated_at
        FROM entity_pages
        WHERE project_id = ?1
        ORDER BY updated_at DESC"
    )?;

    let pages = stmt
        .query_map([&project_id], |row| {
            Ok(EntityPage {
                id: row.get(0)?,
                project_id: row.get(1)?,
                entity_id: row.get(2)?,
                entity_type: row.get(3)?,
                title: row.get(4)?,
                storage_path: row.get(5)?,
                metadata: row.get(6)?,
                created_at: row.get(7)?,
                updated_at: row.get(8)?,
            })
        })?
        .collect::<Result<Vec<_>, _>>()?;

    Ok(pages)
}

#[tauri::command]
pub fn update_entity_page(
    entity_id: String,
    input: UpdateEntityPageInput,
    state: State<AppState>,
) -> CommandResult<EntityPage> {
    let db = state.db.lock().unwrap();
    let now = Utc::now().to_rfc3339();

    let mut updates = Vec::new();
    let mut params: Vec<String> = Vec::new();

    if let Some(title) = input.title {
        updates.push("title = ?");
        params.push(title);
    }
    if let Some(storage_path) = input.storage_path {
        updates.push("storage_path = ?");
        params.push(storage_path);
    }
    if let Some(metadata) = input.metadata {
        updates.push("metadata = ?");
        params.push(metadata);
    }

    updates.push("updated_at = ?");
    params.push(now);
    params.push(entity_id.clone());

    let query = format!(
        "UPDATE entity_pages SET {} WHERE entity_id = ?",
        updates.join(", ")
    );

    let param_refs: Vec<&str> = params.iter().map(|s| s.as_str()).collect();
    db.execute(&query, rusqlite::params_from_iter(param_refs))?;

    drop(db);

    get_entity_page(entity_id, state)?
        .ok_or_else(|| super::CommandError {
            message: "Entity page not found after update".to_string(),
        })
}

#[tauri::command]
pub fn delete_entity_page(
    entity_id: String,
    state: State<AppState>,
) -> CommandResult<()> {
    let db = state.db.lock().unwrap();
    db.execute("DELETE FROM entity_pages WHERE entity_id = ?1", [&entity_id])?;
    Ok(())
}

File: mcp-server/src/tools/entity-pages.js
import { getDatabase } from '../db.js';
import { generateUUID } from '../utils/uuid.js';
import { getCurrentTimestamp, formatToolResponse } from '../utils/formatters.js';
import { readFileSync, writeFileSync, mkdirSync, existsSync, appendFileSync, unlinkSync } from 'fs';
import { join, dirname } from 'path';
import os from 'os';

/**
 * Get storage path for entity page content
 * Must match Tauri app's storage bucket structure: entity-pages/{projectId}/entities/{type}/{id}.md
 */
function getEntityPageStoragePath(projectId, entityType, entityId) {
  const homeDir = os.homedir();
  const basePath = join(homeDir, 'Library', 'Application Support', 'com.arcanum.app', 'storage');
  const entityPath = join(basePath, 'entity-pages', projectId, 'entities', entityType);

  // Create directories if they don't exist
  if (!existsSync(entityPath)) {
    mkdirSync(entityPath, { recursive: true });
  }

  return join(entityPath, `${entityId}.md`);
}

/**
 * Get base storage path for all entity pages
 * Must match Tauri app's entity-pages bucket
 */
function getBaseStoragePath() {
  const homeDir = os.homedir();
  return join(
    homeDir,
    'Library',
    'Application Support',
    'com.arcanum.app',
    'storage',
    'entity-pages'
  );
}

/**
 * Normalize storage path to absolute path
 * Handles both relative paths from DB and absolute paths
 */
function normalizeStoragePath(storagePath) {
  // If already absolute, return as-is
  if (storagePath.startsWith('/') || storagePath.startsWith(os.homedir())) {
    return storagePath;
  }
  // Convert relative path to absolute
  return join(getBaseStoragePath(), storagePath);
}

/**
 * Ensure file exists for an entity page, creating empty file if missing
 * This repairs orphaned DB records
 */
function ensureFileExists(storagePath, autoRepair = true) {
  if (!existsSync(storagePath)) {
    if (autoRepair) {
      // Create parent directories if needed
      const dir = dirname(storagePath);
      if (!existsSync(dir)) {
        mkdirSync(dir, { recursive: true });
      }
      // Create empty file to repair orphaned record
      writeFileSync(storagePath, '', 'utf8');
      return { repaired: true };
    }
    return { missing: true };
  }
  return { exists: true };
}

export const entityPageTools = [
  {
    name: 'create_entity_page',
    description: 'Create an entity page with markdown content for detailed notes and research',
    inputSchema: {
      type: 'object',
      properties: {
        project_id: { type: 'string', description: 'Project UUID' },
        entity_id: {
          type: 'string',
          description: 'Entity UUID (person, event, theory, place, artifact)',
        },
        entity_type: {
          type: 'string',
          description: 'Type of entity',
          enum: ['person', 'event', 'theory', 'place', 'artifact'],
        },
        title: { type: 'string', description: 'Page title' },
        content: { type: 'string', description: 'Markdown content for the page' },
      },
      required: ['project_id', 'entity_id', 'entity_type', 'title', 'content'],
    },
  },
  {
    name: 'get_entity_page',
    description: 'Get an entity page including its full markdown content',
    inputSchema: {
      type: 'object',
      properties: {
        entity_id: { type: 'string', description: 'Entity UUID' },
      },
      required: ['entity_id'],
    },
  },
  {
    name: 'update_entity_page',
    description: 'Update an entity page by replacing or appending content',
    inputSchema: {
      type: 'object',
      properties: {
        entity_id: { type: 'string', description: 'Entity UUID' },
        content: { type: 'string', description: 'New or additional markdown content' },
        mode: {
          type: 'string',
          description: 'Update mode: replace entire content or append to it',
          enum: ['replace', 'append'],
          default: 'replace',
        },
      },
      required: ['entity_id', 'content'],
    },
  },
  {
    name: 'add_entity_note',
    description: 'Add a timestamped note to an entity page (appends with timestamp)',
    inputSchema: {
      type: 'object',
      properties: {
        entity_id: { type: 'string', description: 'Entity UUID' },
        note: { type: 'string', description: 'Note content' },
      },
      required: ['entity_id', 'note'],
    },
  },
  {
    name: 'repair_entity_page',
    description: 'Repair an entity page that has a DB record but missing file (creates empty file)',
    inputSchema: {
      type: 'object',
      properties: {
        entity_id: { type: 'string', description: 'Entity UUID' },
      },
      required: ['entity_id'],
    },
  },
  {
    name: 'link_entity_pages',
    description: 'Create a relationship link between two entity pages',
    inputSchema: {
      type: 'object',
      properties: {
        project_id: { type: 'string', description: 'Project UUID' },
        from_entity_id: { type: 'string', description: 'Source entity UUID' },
        from_entity_type: { type: 'string', description: 'Source entity type' },
        to_entity_id: { type: 'string', description: 'Target entity UUID' },
        to_entity_type: { type: 'string', description: 'Target entity type' },
        relationship_type: { type: 'string', description: 'Type of relationship' },
        verified: {
          type: 'boolean',
          description: 'Is this relationship verified?',
          default: false,
        },
        notes: { type: 'string', description: 'Notes about the relationship' },
      },
      required: [
        'project_id',
        'from_entity_id',
        'from_entity_type',
        'to_entity_id',
        'to_entity_type',
      ],
    },
  },
  {
    name: 'search_entity_pages',
    description: 'Search for entity pages by title or type',
    inputSchema: {
      type: 'object',
      properties: {
        project_id: { type: 'string', description: 'Project UUID' },
        query: { type: 'string', description: 'Search query for title' },
        entity_type: { type: 'string', description: 'Filter by entity type' },
        limit: { type: 'number', description: 'Max results', default: 10 },
      },
      required: ['project_id'],
    },
  },
];

export const entityPageHandlers = {
  create_entity_page: (args) => {
    const db = getDatabase();
    const id = generateUUID();
    const now = getCurrentTimestamp();

    // Generate storage path
    const storagePath = getEntityPageStoragePath(args.project_id, args.entity_type, args.entity_id);

    // Use transaction for atomicity
    const transaction = db.transaction(() => {
      // First, try to insert DB record
      const stmt = db.prepare(`
        INSERT INTO entity_pages (id, project_id, entity_id, entity_type, title, storage_path, metadata, created_at, updated_at)
        VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)
        RETURNING *
      `);

      const page = stmt.get(
        id,
        args.project_id,
        args.entity_id,
        args.entity_type,
        args.title,
        storagePath,
        null,
        now,
        now
      );

      // Then write file - if this fails, transaction will rollback
      try {
        writeFileSync(storagePath, args.content, 'utf8');
      } catch (fileError) {
        // Clean up any partial file
        try {
          if (existsSync(storagePath)) {
            unlinkSync(storagePath);
          }
        } catch (cleanupError) {
          // Ignore cleanup errors
        }
        throw new Error(`Failed to write entity page file: ${fileError.message}`);
      }

      return page;
    });

    try {
      const page = transaction();
      return formatToolResponse(
        `Entity page created successfully:\nID: ${page.id}\nTitle: ${page.title}\nType: ${page.entity_type}\nPath: ${page.storage_path}`
      );
    } catch (error) {
      // If it's a UNIQUE constraint error, provide helpful message
      if (error.message.includes('UNIQUE constraint failed')) {
        throw new Error(
          `Entity page already exists for entity ${args.entity_id}. Use update_entity_page to modify it, or repair_entity_page if the file is missing.`
        );
      }
      throw error;
    }
  },

  get_entity_page: (args) => {
    const db = getDatabase();

    const stmt = db.prepare('SELECT * FROM entity_pages WHERE entity_id = ?');
    const page = stmt.get(args.entity_id);

    if (!page) {
      throw new Error(`Entity page not found for entity: ${args.entity_id}`);
    }

    // Normalize storage path (handle relative paths from DB)
    const absolutePath = normalizeStoragePath(page.storage_path);

    // Read content from file with status reporting
    let content = '';
    let status = '';

    const fileStatus = ensureFileExists(absolutePath, false);

    if (fileStatus.missing) {
      status =
        '\n⚠️  WARNING: Database record exists but backing file is missing!\n' +
        'Use repair_entity_page to create an empty file, then update_entity_page to add content.\n';
    } else {
      content = readFileSync(absolutePath, 'utf8');
    }

    return formatToolResponse(
      `Entity Page: ${page.title}\nType: ${page.entity_type}\nCreated: ${page.created_at}\n` +
        `Last updated: ${page.updated_at}${status}\n\n---\n\n${content}`
    );
  },

  update_entity_page: (args) => {
    const db = getDatabase();

    const stmt = db.prepare('SELECT * FROM entity_pages WHERE entity_id = ?');
    const page = stmt.get(args.entity_id);

    if (!page) {
      throw new Error(`Entity page not found for entity: ${args.entity_id}`);
    }

    // Normalize storage path (handle relative paths from DB)
    const absolutePath = normalizeStoragePath(page.storage_path);

    const mode = args.mode || 'replace';

    // Auto-repair: ensure file exists before writing
    const fileStatus = ensureFileExists(absolutePath, true);
    let repairNote = '';

    if (fileStatus.repaired) {
      repairNote = ' (auto-repaired missing file)';
    }

    if (mode === 'append') {
      // Read existing content if file was just created
      let existingContent = '';
      if (fileStatus.repaired) {
        // File was just created empty, nothing to preserve
      } else {
        existingContent = readFileSync(absolutePath, 'utf8');
      }
      writeFileSync(absolutePath, existingContent + '\n\n' + args.content, 'utf8');
    } else {
      // Replace entire content
      writeFileSync(absolutePath, args.content, 'utf8');
    }

    // Update timestamp
    const now = getCurrentTimestamp();
    const updateStmt = db.prepare('UPDATE entity_pages SET updated_at = ? WHERE entity_id = ?');
    updateStmt.run(now, args.entity_id);

    return formatToolResponse(`Entity page updated successfully (${mode} mode)${repairNote}`);
  },

  add_entity_note: (args) => {
    const db = getDatabase();

    const stmt = db.prepare('SELECT * FROM entity_pages WHERE entity_id = ?');
    const page = stmt.get(args.entity_id);

    if (!page) {
      throw new Error(`Entity page not found for entity: ${args.entity_id}`);
    }

    // Normalize storage path (handle relative paths from DB)
    const absolutePath = normalizeStoragePath(page.storage_path);

    // Auto-repair: ensure file exists before writing
    const fileStatus = ensureFileExists(absolutePath, true);
    let repairNote = '';

    if (fileStatus.repaired) {
      repairNote = ' (auto-repaired missing file)';
    }

    // Create timestamped note
    const now = getCurrentTimestamp();
    const timestamp = new Date(now).toLocaleString('en-US', {
      year: 'numeric',
      month: 'short',
      day: 'numeric',
      hour: '2-digit',
      minute: '2-digit',
    });
    const note = `\n\n---\n\n**Note added ${timestamp}:**\n\n${args.note}`;

    // Read existing content and append
    let existingContent = '';
    if (!fileStatus.repaired) {
      existingContent = readFileSync(absolutePath, 'utf8');
    }
    writeFileSync(absolutePath, existingContent + note, 'utf8');

    // Update timestamp
    const updateStmt = db.prepare('UPDATE entity_pages SET updated_at = ? WHERE entity_id = ?');
    updateStmt.run(now, args.entity_id);

    return formatToolResponse(`Note added to entity page successfully${repairNote}`);
  },

  repair_entity_page: (args) => {
    const db = getDatabase();

    const stmt = db.prepare('SELECT * FROM entity_pages WHERE entity_id = ?');
    const page = stmt.get(args.entity_id);

    if (!page) {
      throw new Error(`Entity page not found for entity: ${args.entity_id}`);
    }

    // Normalize storage path (handle relative paths from DB)
    const absolutePath = normalizeStoragePath(page.storage_path);

    const fileStatus = ensureFileExists(absolutePath, false);

    if (fileStatus.exists) {
      return formatToolResponse(
        `Entity page file already exists at:\n${absolutePath}\n\nNo repair needed.`
      );
    }

    // Create the missing file with recursive directory creation
    ensureFileExists(absolutePath, true);

    // Update timestamp to reflect repair
    const now = getCurrentTimestamp();
    const updateStmt = db.prepare('UPDATE entity_pages SET updated_at = ? WHERE entity_id = ?');
    updateStmt.run(now, args.entity_id);

    return formatToolResponse(
      `Entity page repaired successfully!\n\n` +
        `Created empty file at: ${absolutePath}\n\n` +
        `You can now use update_entity_page or add_entity_note to add content.`
    );
  },

  link_entity_pages: (args) => {
    const db = getDatabase();
    const id = generateUUID();
    const now = getCurrentTimestamp();

    const stmt = db.prepare(`
      INSERT INTO entity_links (
        id, project_id, from_entity_id, from_entity_type,
        to_entity_id, to_entity_type, relationship_type,
        verified, notes, created_at
      )
      VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
      RETURNING *
    `);

    const link = stmt.get(
      id,
      args.project_id,
      args.from_entity_id,
      args.from_entity_type,
      args.to_entity_id,
      args.to_entity_type,
      args.relationship_type || null,
      args.verified ? 1 : 0,
      args.notes || null,
      now
    );

    return formatToolResponse(
      `Entity pages linked successfully:\n${args.from_entity_type} → ${args.to_entity_type}\nRelationship: ${link.relationship_type || 'related'}`
    );
  },

  search_entity_pages: (args) => {
    const db = getDatabase();
    const limit = args.limit || 10;

    let query = 'SELECT * FROM entity_pages WHERE project_id = ?';
    const params = [args.project_id];

    if (args.query) {
      query += ' AND title LIKE ? COLLATE NOCASE';
      params.push(`%${args.query}%`);
    }

    if (args.entity_type) {
      query += ' AND entity_type = ?';
      params.push(args.entity_type);
    }

    query += ' ORDER BY updated_at DESC LIMIT ?';
    params.push(limit);

    const stmt = db.prepare(query);
    const pages = stmt.all(...params);

    // Check which pages have missing files
    const pagesWithStatus = pages.map((page) => {
      const absolutePath = normalizeStoragePath(page.storage_path);
      const hasFile = existsSync(absolutePath);
      return {
        ...page,
        file_status: hasFile ? 'ok' : 'MISSING',
      };
    });

    const missingCount = pagesWithStatus.filter((p) => p.file_status === 'MISSING').length;
    let warning = '';

    if (missingCount > 0) {
      warning =
        `\n\n⚠️  WARNING: ${missingCount} page(s) have missing backing files!\n` +
        `Use repair_entity_page to fix them.\n`;
    }

    return formatToolResponse(
      `Found ${pages.length} entity pages:${warning}\n${JSON.stringify(pagesWithStatus, null, 2)}`
    );
  },
};
